tests/unit/pmodel/test_optimal_chi.py

```diff
@@ -0,0 +1,182 @@
+"""Testing CalcOptimal submodule."""
+
+
+import numpy as np
+import pytest
+
+from pyrealm.pmodel.optimal_chi import (
+    OptimalChiC4,
+    OptimalChiC4NoGamma,
+    OptimalChiC4NoGammaRootzoneStress,
+    OptimalChiC4RootzoneStress,
+    OptimalChiLavergne20C3,
+    OptimalChiLavergne20C4,
+    OptimalChiPrentice14,
+    OptimalChiPrentice14RootzoneStress,
+)
+from pyrealm.pmodel.pmodel_environment import PModelEnvironment
+
+
+@pytest.fixture
+def photo_env():
+    """Photosynthesis Environment setup."""
+    return PModelEnvironment(
+        tc=np.array([20]),
+        vpd=np.array([1000]),
+        co2=np.array([400]),
+        patm=np.array([101325.0]),
+        rootzonestress=np.array([1]),
+        theta=np.array([0.5]),
+    )
+
+
+@pytest.mark.parametrize(
+    """optimal_chi_class""",
+    [
+        OptimalChiPrentice14,
+        OptimalChiPrentice14RootzoneStress,
+        OptimalChiC4NoGammaRootzoneStress,
+        OptimalChiC4,
+        OptimalChiC4RootzoneStress,
+        OptimalChiLavergne20C3,
+        OptimalChiLavergne20C4,
+        OptimalChiC4NoGamma,
+    ],
+)
+def test_set_beta(optimal_chi_class, photo_env):
+    """Test that beta is set correctly."""
+    optimal_chi_instance = optimal_chi_class(env=photo_env)
+    optimal_chi_instance.set_beta()
+    # Test that beta attribute is set correctly
+    assert isinstance(optimal_chi_instance.beta, np.ndarray)
+
+
+@pytest.mark.parametrize(
+    """optimal_chi_class""",
+    [
+        OptimalChiPrentice14,
+        OptimalChiPrentice14RootzoneStress,
+        OptimalChiC4NoGammaRootzoneStress,
+        OptimalChiC4,
+        OptimalChiC4RootzoneStress,
+        OptimalChiLavergne20C3,
+        OptimalChiLavergne20C4,
+        OptimalChiC4NoGamma,
+    ],
+)
+def test_estimate_chi(optimal_chi_class, photo_env):
+    """Test that chi is estimated correctly."""
+    optimal_chi_instance = optimal_chi_class(env=photo_env)
+    optimal_chi_instance.set_beta()
+    optimal_chi_instance.estimate_chi()
+    # Test that chi and other related attributes are calculated correctly
+    assert isinstance(optimal_chi_instance.chi, np.ndarray)
+    assert isinstance(optimal_chi_instance.mc, np.ndarray)
+    assert isinstance(optimal_chi_instance.mj, np.ndarray)
+    assert isinstance(optimal_chi_instance.mjoc, np.ndarray)
+
+
+@pytest.mark.parametrize(
+    argnames=["subclass", "pmodelenv_args", "expected"],
+    argvalues=[
+        (
+            OptimalChiPrentice14,
+            dict(tc=20, patm=101325, co2=400, vpd=1000),
+            dict(chi=0.69435, mc=0.33408, mj=0.7123, mjoc=2.13211),
+        ),
+        (
+            OptimalChiPrentice14RootzoneStress,
+            dict(tc=20, patm=101325, co2=400, vpd=1000, rootzonestress=0.5),
```

Comment:
Some rough bounds on these input variables given in the bounds checking here: https://github.com/ImperialCollegeLondon/pyrealm/blob/develop/pyrealm/pmodel/pmodel_environment.py#L91-L99 (and below for theta and rootzonestress).

As we discussed, the issue is that remotely sensed inputs have all sorts of odd values in them, so `pyrealm` is permissive about allowing extreme values but does issue warnings in an attempt to alert users. There are some more details here: https://pyrealm.readthedocs.io/en/latest/users/pmodel/pmodel_details/extreme_values.html#realistic-input-values

More concretely:
* `co2` and `patm` _cannot_ be negative. 
* `theta` and `rootzonestress` must be in `[0,1]`.

In the outputs:
* `beta` and `xi` must be positive.
* `mc` and `mj` must be in `[0,1]` and hence `mjoc = mj/mc` must be positive.
* `ci` must be positive and `chi` must be in `[0,1]`.

I'm vaguely wondering whether `pyrealm` should add a second layer of value checking to eliminate _impossible_ inputs. So `bounds_checker` might implement hard and soft bounds, where exceeding the hard bound is physically impossible but exceeding the soft bounds is just a bit unlikely.

Reply:
Let me know if you need more details - sorry for the delay in getting this to you.

Reply:
@davidorme As I checked,
 - `bounds_checker` does issue warning for out-of-bound values and 
 - `bounds_mask` replaces the out-of-bound values with `np.nan`
  
But how these nan values are further handled is not clear. When I included the the testcase including nan values, the unit test for optimalchi failed. So maybe we can look for an appropriate method of handling the nan values as they may create issues in calculations in future.

Reply:
I am writing the testcase to check the performance of `bounds_checker` for `Optimal_chi`. That could possibly guide us on whether the second layer of bounds_check is required.